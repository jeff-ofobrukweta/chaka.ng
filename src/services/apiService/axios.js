import axios from "axios";

const instance = axios.create({
    // baseURL: `https://c89940da-8733-4b38-9260-058a90e8895f.mock.pstmn.io`,
    baseURL: `https://test-api.chaka.io`,
    crossdomain: true,
    headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
        Authorization: {
            toString() {
                return `Bearer ${localStorage.getItem("AUTH_TOKEN")}`;
                // return `Bearer "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VyIjp7ImlkIjoxLCJjaGFrYUlEIjoxMDEyNTY3ODEwLCJlbWFpbCI6InNldW5AY2hha2EubmciLCJyb2xlIjoiQ1VTVE9NRVIifSwiaWF0IjoxNTYxMzgzNzI3LCJleHAiOjE1NjEzODM3NDJ9.awx6JB7PHppMtP-PdsylXCaVqqcZRrR6SoeL-8gEohc"`;
                
            }
        }
    }
});

export default instance;
