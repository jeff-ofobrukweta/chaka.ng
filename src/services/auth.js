import store from "../store/index";

/**
 * Resets state and calls logout mutation
 */
function resetState() {
    // store.commit("LOGOUT");
    store.commit("SET_LOGGED_USER", {});
    // store.commit("RESET_ALL");
    store.commit("SET_LOGGED_IN", false);
    localStorage.removeItem("AUTH_TOKEN");
    localStorage.removeItem("REFRESH_TOKEN");
    return true;
}

/**
 * retrieves the current user session from state
 */
export function getSession() {
    return store.getters.getLoggedUser;
}

/**
 * Checks for loggedIn status
 * @returns {Boolean} loggedIn state
 */
export function isLoggedIn() {
    return store.getters.isLoggedIn;
}

/**
 * Auth guard that allows non-auhtenticated users only.
 * @param to - next route
 * @param from - previous route
 * @param next - callback to transfer control to the next middleware
 */
export function noAuthOnly(to, from, next) {
    if (to.name === "login" && from.path.startsWith("/dashboard")) {
        resetState();
        next();
    } else if (isLoggedIn()) {
        if (to.name === "calculator") {
            next("/dashboard/calculator");
        } else if (to.name === "gifts") {
            //  TO-DO:: Put back when Gifts is ready
            // next("/dashboard/gifts");
            next("/login");
        } else if (to.name === "valentine" || to.name === "valentine-result") {
            resetState();
            next();
        } else next("/dashboard");
    } else {
        next();
    }
}

/**
 * Auth guard allows authenticated users only.
 * @param to - next route
 * @param from - previous route
 * @param next - callback to transfer control to the next middleware
 */
export function requireAuth(to, from, next) {
    if (isLoggedIn()) {
        next();
    } else {
        next({ name: "login" });
    }
}

/**
 * clears the current session
 */
export function clearSession(to, from, next) {
    next({ name: "login" });
}
