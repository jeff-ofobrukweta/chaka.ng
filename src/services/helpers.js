export const getRandomInteger = maximum => Math.floor(Math.random() * Math.floor(maximum));
