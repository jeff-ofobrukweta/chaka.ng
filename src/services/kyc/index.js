const kycFields = [
    {
        name: "Disclosure Name",
        value: "disclosureName",
        type: "text"
    },
    {
        name: "Address",
        value: "address",
        type: "text"
    },
    {
        name: "Address Proof",
        value: "addressProofUrl",
        type: "image"
    },
    {
        name: "Annual Income",
        value: "annualIncome",
        type: "button",
        options: [
            {
                name: "< ₦500k",
                value: "LESS_THAN_500K"
            },
            {
                name: "₦500k - ₦5M",
                value: "500K_5MILLION"
            },
            {
                name: "₦5M+",
                value: "GREATER_THAN_5MILLION"
            }
        ]
    },
    {
        name: "Bank Name",
        value: "bankCode",
        type: "select"
    },
    {
        name: "Account Number",
        value: "bankAcctNo",
        type: "number"
    },
    {
        name: "BVN",
        value: "bvn",
        type: "number"
    },
    {
        name: "NIN",
        value: "nin",
        type: "number"
    },
    {
        name: "Country",
        value: "country",
        type: "select"
    },
    {
        name: "CSCS Number",
        value: "cscsNumber",
        type: "number"
    },
    {
        name: "Are you a director Of Public Co.",
        value: "directorOfPublicCo",
        type: "button",
        options: [
            {
                name: "Yes",
                value: true
            },
            {
                name: "No",
                value: false
            }
        ]
    },
    {
        name: "Employed By Broker",
        value: "employedByBroker",
        type: "button",
        options: [
            {
                name: "Yes",
                value: true
            },
            {
                name: "No",
                value: false
            }
        ]
    },
    {
        name: "Employment Company",
        value: "employmentCompany",
        type: "text"
    },
    {
        name: "Employment Position",
        value: "employmentPosition",
        type: "select"
    },
    {
        name: "Employment Status",
        value: "employmentStatus",
        type: "button",
        options: [
            {
                name: "Employed",
                value: "EMPLOYED"
            },
            {
                name: "Self-Employed",
                value: "SELF_EMPLOYED"
            },
            {
                name: "Unemployed",
                value: "UNEMPLOYED"
            }
        ]
    },
    {
        name: "Employment Type",
        value: "employmentType",
        type: "select"
    },
    {
        name: "Gender",
        value: "gender",
        type: "button",
        options: [
            {
                name: "Male",
                value: "male"
            },
            {
                name: "Female",
                value: "female"
            }
        ]
    },
    {
        name: "National ID/Passport",
        value: "idPhotoUrl",
        type: "image"
    },
    {
        name: "Investment Experience",
        value: "investmentExperience",
        type: "button",
        options: [
            {
                name: "0 - 2 Yrs",
                value: "YRS_0_2"
            },
            {
                name: "2 - 5 Yrs",
                value: "YRS_3_5"
            },
            {
                name: "5 Yrs+",
                value: "YRS_5_plus"
            }
        ]
    },
    {
        name: "Investment Objectives",
        value: "investmentObjectives",
        type: "button",
        options: [
            {
                name: "Protection",
                value: "PROTECTION"
            },
            {
                name: "Growth",
                value: "GROWTH"
            },
            {
                name: "Income",
                value: "INCOME"
            }
        ]
    },
    {
        name: "LGA",
        value: "lg",
        type: "select"
    },
    // {
    //     name: "Maiden Name",
    //     value: "maidenName",
    //     type: "text"
    // },
    {
        name: "Net Worth Liquid",
        value: "networthLiquid",
        type: "button",
        options: [
            {
                name: "< ₦5M",
                value: "<N5m"
            },
            {
                name: "₦5M - ₦50M",
                value: "N5m-N50m"
            },
            {
                name: "₦50M+",
                value: "N50m+"
            }
        ]
    },
    {
        name: "Net Worth Total",
        value: "networthTotal",
        type: "button",
        options: [
            {
                name: "< ₦5M",
                value: "<N5m"
            },
            {
                name: "₦5M - ₦50M",
                value: "N5m-N50m"
            },
            {
                name: "₦50M+",
                value: "N50m+"
            }
        ]
    },
    // {
    //     name: "Next of Kin Phone",
    //     value: "nextOfKinPhone",
    //     type: "text"
    // },
    {
        name: "Passport Image",
        value: "passportUrl",
        type: "image"
    },
    {
        name: "Are you Politically Exposed?",
        value: "pepStatus",
        type: "button",
        options: [
            {
                name: "Yes",
                value: true
            },
            {
                name: "No",
                value: false
            }
        ]
    },
    {
        name: "Name of Politically Exposed",
        value: "pepNames",
        type: "text"
    },
    {
        name: "Phone",
        value: "phone",
        type: "number"
    },
    {
        name: "Risk Tolerance",
        value: "riskTolerance",
        type: "button",
        options: [
            {
                name: "Conservative",
                value: "CONSERVATIVE"
            },
            {
                name: "Neutral",
                value: "NEUTRAL"
            },
            {
                name: "Risk Seeking",
                value: "RISK-SEEKING"
            }
        ]
    }
];

export default kycFields;
