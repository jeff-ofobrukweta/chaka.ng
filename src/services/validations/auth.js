const auth = {
    email: value => {
        var re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
        return re.test(String(value).toLowerCase());
    },
    password: value => {
        var re = /(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*\W)/;
        return re.test(String(value).toLowerCase());
    }
};

export default auth;
