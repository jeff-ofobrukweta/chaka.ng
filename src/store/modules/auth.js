import api from "../../services/apiService/api";
import errorFn from "../../services/apiService/error";
import jwtDecode from "jwt-decode";

const state = {
    loggedUser: {}
};

const getters = {
    getLoggedUser: state => state.loggedUser,
    isLoggedIn: state => {
        if (Object.keys(state.loggedUser).length > 0) return true;
        return false;
    }
};

const mutations = {
    SET_LOGGED_USER(state, payload) {
        state.loggedUser = payload;
    }
};

const actions = {
    REGISTER: ({ commit }, payload) => {
        commit("RESET_REQ", null, { root: true });
        commit("REQ_INIT", null, { root: true });
        return new Promise((resolve, reject) => {
            return api.post("/auth/signup", payload).then(
                resp => {
                    if (resp.status === 200) {
                        commit("REQ_SUCCESS", null, { root: true });
                        resolve(true);
                    } else {
                        errorFn(resp, "register");
                        resolve(false);
                    }
                },
                error => {
                    errorFn(error.response, "register");
                    resolve(false);
                }
            );
        });
    },
    LOGIN: ({ commit, dispatch }, payload) => {
        commit("RESET_REQ", null, { root: true });
        commit("REQ_INIT", null, { root: true });
        return new Promise((resolve, reject) => {
            return api.post(`/auth/login`, payload).then(
                resp => {
                    if (resp.status === 200) {
                        commit("REQ_SUCCESS", null, { root: true });
                        localStorage.setItem("AUTH_TOKEN", resp.data.data.token);
                        dispatch("GET_LOGGED_USER");
                        resolve(true);
                    } else {
                        errorFn(resp, "login");
                        resolve(false);
                    }
                },
                error => {
                    errorFn(error.response, "login");
                    resolve(false);
                }
            );
        });
    },
    GET_LOGGED_USER: ({ commit }) => {
        const token = jwtDecode(localStorage.getItem("AUTH_TOKEN"));
        commit("RESET_REQ", null, { root: true });
        commit("REQ_INIT", null, { root: true });
        return new Promise((resolve, reject) => {
            return api.get(`/users/${token.user.chakaID}`).then(
                resp => {
                    if (resp.status === 200) {
                        commit("REQ_SUCCESS", null, { root: true });
                        commit("REQ_SUCCESS", null, { root: true });
                        commit("SET_LOGGED_USER", resp.data.data.user);
                        resolve(true);
                    } else {
                        errorFn(resp, "login");
                        resolve(false);
                    }
                },
                error => {
                    errorFn(error.response, "login");
                    resolve(false);
                }
            );
        });
    }
};

export default {
    state,
    actions,
    getters,
    mutations
};
