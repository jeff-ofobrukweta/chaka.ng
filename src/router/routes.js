import { clearSession, requireAuth, noAuthOnly } from "../services/auth";

const Home = () => import(/* webpackChunkName: "home" */ "../views/Home");
const VerificationSent = () =>
    import(/* webpackChunkName: "verification-sent" */ "../views/VerificationSent");
const ForgotPassword = () =>
    import(/* webpackChunkName: "forgot-password" */ "../views/auth/ForgotPassword");
const ResetPassword = () =>
    import(/* webpackChunkName: "reset-password" */ "../views/auth/ResetPassword");
const ConfirmEmail = () =>
    import(/* webpackChunkName: "confirm-email" */ "../views/auth/ConfirmEmail");
const Login = () => import(/* webpackChunkName: "login" */ "../views/auth/Login");
const Policies = () => import(/* webpackChunkName: "policies" */ "../views/Policies");
const Developers = () => import(/* webpackChunkName: "developers" */ "../views/Developers");
const Privacy = () => import(/* webpackChunkName: "privacy" */ "../views/Privacy");
const Calculator = () => import(/* webpackChunkName: "calculator" */ "../views/Calculator");
const Terms = () => import(/* webpackChunkName: "terms" */ "../views/Terms");
const About = () => import(/* webpackChunkName: "about" */ "../views/About");
const Faq = () => import(/* webpackChunkName: "faq" */ "../views/Faq");
const Valentine = () => import(/* webpackChunkName: "valentine" */ "../views/Valentine");
const Gifts = () => import(/* webpackChunkName: "gifts" */ "../views/Gifts");
const Logout = () => import(/* webpackChunkName: "logout" */ "../views/auth/Logout");
const Register = () => import(/* webpackChunkName: "register" */ "../views/auth/Register");
const Disclosures = () => import(/* webpackChunkName: "disclosures" */ "../views/Disclosure");
const AppHome = () => import(/* webpackChunkName: "app-home" */ "../views/AppHome");
const DashboardLayout = () =>
    import(/* webpackChunkName: "dashboard" */ "../layouts/DashboardLayout");
const Portfolio = () => import(/* webpackChunkName: "portfolio" */ "../views/dashboard/Portfolio");
const PortfolioDetails = () =>
    import(/* webpackChunkName: "portfolio-details" */ "../views/dashboard/PortfolioDetails");
const Explore = () => import(/* webpackChunkName: "explore" */ "../views/dashboard/Explore");
const ExploreCollection = () =>
    import(/* webpackChunkName: "explore-collection" */ "../views/dashboard/SingleCollection");
const Categories = () =>
    import(/* webpackChunkName: "categories" */ "../views/dashboard/Categories");
const DashboardCalculator = () =>
    import(/* webpackChunkName: "dashboard-calculator" */ "../views/dashboard/Calculator");
const DashboardGifts = () =>
    import(/* webpackChunkName: "dashboard-gifts" */ "../views/dashboard/Gifts");
const AccountsIndex = () =>
    import(/* webpackChunkName: "accounts-index" */ "../views/dashboard/accounts/Index");
const AccountsOverview = () =>
    import(/* webpackChunkName: "accounts-overview" */ "../views/dashboard/accounts/Overview");
const AccountsWallet = () =>
    import(/* webpackChunkName: "accounts-wallet" */ "../views/dashboard/accounts/Wallet");
const AccountsStatements = () =>
    import(/* webpackChunkName: "accounts-statements" */ "../views/dashboard/accounts/Statements");
const AccountsHistory = () =>
    import(/* webpackChunkName: "accounts-history" */ "../views/dashboard/accounts/History");
const AccountsSettings = () =>
    import(/* webpackChunkName: "accounts-settings" */ "../views/dashboard/accounts/Settings");
const AccountStamps = () =>
    import(/* webpackChunkName: "accounts-stamps" */ "../views/dashboard/accounts/Stamps");
const Singlestock = () =>
    import(/* webpackChunkName: "accounts-stamps" */ "../views/dashboard/Singlestock");

const routes = [
    {
        path: "/",
        name: "home",
        component: Home,
        beforeEnter: noAuthOnly
    },
    {
        path: "/relationship-calculator",
        name: "valentine",
        component: Valentine
    },
    {
        path: "/get-the-app",
        name: "apphome",
        component: AppHome
    },
    {
        path: "/about",
        name: "about",
        component: About,
        beforeEnter: noAuthOnly
    },
    {
        path: "/calculator",
        name: "calculator",
        component: Calculator,
        beforeEnter: noAuthOnly
    },
    {
        path: "/gifts",
        name: "gifts",
        component: Gifts,
        beforeEnter: noAuthOnly
    },
    {
        path: "/auth/reset-password/:chakaID/:token",
        name: "reset-password",
        component: ResetPassword
    },
    {
        path: "/auth/confirm-email-token/:chakaID/:token",
        name: "confirm-email",
        component: ConfirmEmail
    },
    {
        path: "/policies",
        name: "policies",
        component: Policies,
        beforeEnter: noAuthOnly
    },
    {
        path: "/developers",
        name: "developers",
        component: Developers,
        beforeEnter: noAuthOnly
    },
    {
        path: "/privacy",
        name: "privacy",
        component: Privacy
    },
    {
        path: "/faq",
        name: "faq",
        component: Faq,
        beforeEnter: noAuthOnly
    },
    {
        path: "/terms",
        name: "terms",
        component: Terms,
        beforeEnter: noAuthOnly
    },
    {
        path: "/verification-sent",
        name: "verification-sent",
        component: VerificationSent,
        beforeEnter: noAuthOnly
    },
    {
        path: "/disclosures",
        name: "disclosures",
        component: Disclosures,
        beforeEnter: noAuthOnly
    },
    {
        path: "/forgot-password",
        name: "forgot-password",
        component: ForgotPassword,
        beforeEnter: noAuthOnly
    },
    {
        path: "/login",
        name: "login",
        component: Login,
        beforeEnter: noAuthOnly
    },
    {
        path: "/logout",
        name: "logout",
        component: Logout,
        beforeEnter: clearSession
    },
    {
        path: "/register",
        name: "register",
        component: Register,
        beforeEnter: noAuthOnly
    },
    {
        path: "/dashboard",
        name: "dashboard",
        component: DashboardLayout,
        redirect: "/dashboard/portfolio",
        beforeEnter: requireAuth,
        children: [
            {
                path: "portfolio",
                name: "portfolio",
                component: Portfolio
            },
            {
                path: "portfolio/details/:type",
                name: "portfolio-details",
                component: PortfolioDetails
            },
            {
                path: "explore",
                name: "explore",
                component: Explore
            },
            {
                path: "collection/:name",
                name: "explore-collection",
                component: ExploreCollection
            },
            {
                path: "calculator",
                name: "dashboard-calculator",
                component: DashboardCalculator
            },
            {
                path: "gifts",
                name: "dashboard-gifts",
                component: DashboardGifts
            },
            {
                path: "categories/:category",
                name: "categories",
                component: Categories
            },
            {
                path: "singlestock/:symbol",
                name: "singlestock",
                component: Singlestock
            },
            {
                path: "accounts",
                component: AccountsIndex,
                redirect: "/dashboard/accounts/overview",
                children: [
                    {
                        path: "overview",
                        name: "accounts-overview",
                        component: AccountsOverview
                    },
                    {
                        path: "wallet",
                        name: "accounts-wallet",
                        component: AccountsWallet
                    },
                    {
                        path: "statements",
                        name: "accounts-statements",
                        component: AccountsStatements
                    },
                    {
                        path: "history",
                        name: "accounts-history",
                        component: AccountsHistory
                    },
                    {
                        path: "settings",
                        name: "accounts-settings",
                        component: AccountsSettings
                    },
                    {
                        path: "ownerships",
                        name: "accounts-stamps",
                        component: AccountStamps
                    }
                ]
            }
        ]
    },
    {
        path: "*",
        redirect: "/dashboard"
    }
];
export default routes;
