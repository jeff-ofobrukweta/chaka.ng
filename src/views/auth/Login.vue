<template>
    <main class="main" role="main">
        <div class="auth-box">
            <form class="auth-form" @submit.prevent="login" novalidate>
                <h3 class="auth-form__title">Login to Chaka</h3>
                <p class="auth-form__subtitle">Enter your login details to proceed.</p>
                <div class="auth-form__group">
                    <label class="form__label"
                        >Email
                        <form-input
                            type="email"
                            name="email"
                            v-model="itemData.email"
                            :error-message="errors.email"
                            placeholder="Email Address"
                            @reset="resetError"
                    /></label>
                </div>
                <div class="auth-form__group">
                    <label class="form__label"
                        >Password
                        <form-input
                            type="password"
                            name="password"
                            v-model="itemData.password"
                            placeholder="Password"
                            :error-message="errors.password"
                            @reset="resetError"
                    /></label>
                </div>
                <error-block type="login" />
                <div class="auth-form__group">
                    <div>
                        <action-button
                            type="submit"
                            :disabled="!formValid"
                            :pending="loading"
                            :classes="['btn-full', 'btn__primary']"
                            >Login</action-button
                        >
                    </div>
                    <section class="auth-form__meta">
                        <p>
                            Don't have an account?
                            <router-link class="primary" :to="{ name: 'register' }"
                                >Register now</router-link
                            >
                        </p>
                        <router-link class="primary" :to="{ name: 'forgot-password' }"
                            >Forgot Password</router-link
                        >
                    </section>
                </div>
            </form>
        </div>
    </main>
</template>

<script>
import { mapActions, mapMutations } from "vuex";
import auth from "../../services/validations/auth";
import store from "../../store/index";

export default {
    name: "Login",
    data() {
        return {
            itemData: {},
            errors: {},
            loading: false
        };
    },
    computed: {
        formValid() {
            if (this.loading || Object.keys(this.errors).length > 0) return false;
            return true;
        }
    },
    methods: {
        ...mapActions(["LOGIN"]),
        ...mapMutations(["RESET_REQ", "RESET_ALL"]),
        login() {
            this.RESET_REQ();
            if (!this.itemData.email) {
                this.$set(this.errors, "email", "Field is required");
            } else if (!auth.email(this.itemData.email)) {
                this.$set(this.errors, "email", "Invalid email");
            }
            if (!this.itemData.password) {
                this.$set(this.errors, "password", "Field is required");
            }
            // const p = auth.password(this.itemData.password);
            // console.log(p);
            // else if(){
            //     console.log()
            // }
            if (Object.keys(this.errors).length > 0) {
                return false;
            }
            this.loading = true;
            this.LOGIN(this.itemData).then(resp => {
                this.loading = false;
                if (resp) this.$router.push({ name: "dashboard" });
            });
        },
        resetError() {
            this.errors = {};
        }
    },
    mounted() {
        this.resetError();
        document.title = "Chaka - Login";
        document.getElementsByTagName("meta").keywords.content =
            "nigerian stock exchange, US stock market, nigeria stock market, online investment, investing, capital market, stock trading, stockbroker, stocks, shares, investment passport, chaka, nse, nyse";
        document.getElementsByTagName("meta").description.content =
            "Invest and Trade thousands of companies across 40+ countries through the Nigerian and US Stock Exchanges. Regulated in both Nigeria and the US by Securities Exchange Commission, FINRA, IRS and SIPC.";
        this.RESET_ALL();
    }
};
</script>
