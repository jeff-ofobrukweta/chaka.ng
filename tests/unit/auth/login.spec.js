import { shallowMount, createLocalVue, mount } from '@vue/test-utils';
import Vuex from 'vuex';
import Router from 'vue-router';
import axios from 'axios';

import sinon from 'sinon';
import { shallow } from 'enzyme';
import PubSub from 'pubsub-js';
import routes from '@/router/routes';
import Login from '@/views/auth/Login';
import { auth, API } from '../services';
import { login } from '../services/auth';


const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(Router);

const router = new Router({ routes });

describe('Components on Login Page', () => {
    let mutations;
    let store;
    let wrapper;
    beforeEach(() => {
        mutations = {
            RESET_ALL: jest.fn()
        };
        store = new Vuex.Store({
            state: {},
            mutations
        });

        wrapper = mount(Login, {
            stubs: ['form-input', 'error-block', 'action-button'],
            localVue,
            router,
            store
        });
    });

    it('is a vue instance', () => {
        expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it('matches snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('calls RESET_REQ mutation on mounted', () => {
        expect(mutations.RESET_ALL).toHaveBeenCalled();
    });

    it('should have 2 input fields', () => {
        expect(wrapper.findAll('form-input-stub').length).toEqual(2);
    });

    it('should have one email input field', () => {
        expect(wrapper.findAll('form-input-stub[type="email"]').length).toEqual(1);
    });

    it('should have one password input field', () => {
        expect(wrapper.findAll('form-input-stub[type="password"]').length).toEqual(1);
    });

    it('should have one button', () => {
        expect(wrapper.findAll('action-button-stub').length).toEqual(1);
    });

    it('should have a link to the register page', async () => {
        wrapper.find({ ref: 'register-link' }).trigger('click');
        await wrapper.vm.$nextTick();
        expect(wrapper.vm.$route.name).toBe('register');
    });

    it('should have a link to the forgot password page', async () => {
        wrapper.find({ ref: 'forgot-link' }).trigger('click');
        await wrapper.vm.$nextTick();
        expect(wrapper.vm.$route.name).toBe('forgot-password');
    });

    // it('should make an API to the login endpoint', async () => {
    //     jest.mock('axios')
    //     const data = {
    //         data: {
    //             hits: [
    //                 {
    //                     objectID: '1',
    //                     title: 'a',
    //                 },
    //                 {
    //                     objectID: '2',
    //                     title: 'b',
    //                 },
    //             ],
    //         },
    //     };
    //     axios.mockResolvedValue()
    //     axios.post.mockImplementationOnce(() => Promise.resolve(data))
    //     await expect(auth({ email: 'dev@chaka.ng', password: 'Pa55w0rd' })).resolves.toEqual(data);
    //     expect(axios.post).toHaveBeenCalledWith(
    //         `${API}/auth/login`,
    //     );
    // })
});

describe('API calls on Login Page', () => {
    let mutations;
    let store;
    let wrapper;
    beforeEach(() => {
        mutations = {
            RESET_ALL: sinon.spy()
        };
        store = new Vuex.Store({
            state: {},
            mutations
        });
        wrapper = shallowMount(Login, {
            stubs: ['form-input', 'error-block', 'action-button'],
            localVue,
            router,
            store
        });
    });
    afterEach(() => {
        sinon.restore();
    });
    it('should test all subscribers', () => {
        const message = 'sample message';
        const stub = sinon.stub().throws();
        const spy1 = sinon.spy();
        const spy2 = sinon.spy();

        PubSub.subscribe(message, stub);
        PubSub.subscribe(message, spy1);
        PubSub.subscribe(message, spy2);

        PubSub.publishSync(message, undefined);
        expect(spy1.called);
        expect(spy2.called);
        expect(stub.calledBefore(spy1));
    });
    it('should make an API call when login button clicked', async () => {
        const credentials = {
            email: 'dev@chaka.ng',
            password: 'Pa55w0rd'
        };

        const url = '/users';

        const post = sinon.stub(axios, 'post');
        post.yields();

        const callback = sinon.spy();
        wrapper.find('action-button-stub').trigger('click');
        await wrapper.vm.$nextTick();
        login(credentials, callback);
        post.restore();
        sinon.assert.calledWith(post, url, credentials);
    });
    it('should should contain the right credentials when login button clicked', async () => {
        const credentials = {
            email: 'dev@chaka.ng',
            password: 'Pa55w0rd'
        };

        const url = '/users';

        const post = sinon.stub(axios, 'post');
        post.yields();

        const callback = sinon.spy();
        wrapper.find('action-button-stub').trigger('click');
        await wrapper.vm.$nextTick();
        login(credentials, callback);
        post.restore();
        sinon.assert.calledWith(post, url, credentials);
    });
});
