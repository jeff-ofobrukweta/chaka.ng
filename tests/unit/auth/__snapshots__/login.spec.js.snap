// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components on Login Page matches snapshot 1`] = `
<main role="main" class="main">
  <div class="auth-box">
    <form novalidate="novalidate" class="auth-form">
      <h3 class="auth-form__title">Login to Chaka</h3>
      <p class="auth-form__subtitle">Enter your login details to proceed.</p>
      <div class="auth-form__group"><label class="form__label">Email <form-input-stub type="email" name="email" placeholder="Email Address"></form-input-stub></label></div>
      <div class="auth-form__group"><label class="form__label">Password
          <form-input-stub type="password" name="password" placeholder="Password"></form-input-stub></label></div>
      <error-block-stub type="login"></error-block-stub>
      <div class="auth-form__group">
        <div style="margin: 1rem 0px;">
          <action-button-stub type="submit" classes="btn-full,btn__primary">Login</action-button-stub>
        </div>
        <div class="or-separator_flex__6ViaO">
          <hr class="or-separator_line__x5QkI">
          <div class="or-separator_text__2YQdZ">or</div>
          <hr class="or-separator_line__x5QkI">
        </div>
        <div class="socials_container" style="cursor: pointer;">
          <section class="socials_btn">
            <div class="_provider-button_label__1gtgk"><span style="white-space: nowrap;">Sign in with</span> <svg viewBox="5 -5 30 30" enable-background="new 5 -5 30 30" title="google" class="google_img">
                <path fill="#fff" d="M15.3-4.2C11.6-3 8.4-.2 6.6 3.2 6 4.5 5.6 5.7 5.3 7c-.7 3.3-.2 6.7 1.3 9.7 1 1.9 2.4 3.7 4.2 5 1.6 1.3 3.5 2.2 5.6 2.7 2.6.7 5.3.7 7.8.1 2.3-.5 4.5-1.6 6.3-3.2 1.9-1.7 3.2-3.9 3.9-6.2.8-2.6.9-5.3.4-8-4.8 0-9.6 0-14.4 0 0 2 0 3.9 0 5.9 2.8 0 5.6 0 8.3 0-.3 1.9-1.5 3.6-3.1 4.6-1 .7-2.2 1.1-3.4 1.3-1.2.2-2.5.2-3.7 0-1.2-.2-2.4-.7-3.4-1.4-1.6-1.1-2.9-2.8-3.5-4.6-.7-1.9-.7-4 0-5.8.5-1.3 1.2-2.5 2.2-3.5 1.2-1.2 2.8-2.1 4.6-2.5 1.5-.3 3-.2 4.5.2 1.2.4 2.4 1 3.3 1.9.9-.9 1.9-1.8 2.8-2.8.5-.5 1-1 1.5-1.5-1.4-1.3-3.1-2.3-4.9-3-3.3-1.2-7-1.2-10.3-.1z"></path>
                <path fill="#EA4335" d="M15.3-4.2c3.3-1.1 7-1.1 10.3.1 1.8.7 3.5 1.7 4.9 3-.5.5-1 1-1.5 1.5-.9.9-1.9 1.8-2.8 2.8-.9-.9-2.1-1.5-3.3-1.9-1.4-.4-3-.5-4.5-.2-1.7.4-3.3 1.2-4.6 2.5-1 1-1.8 2.2-2.2 3.5-1.7-1.3-3.3-2.5-5-3.8 1.8-3.5 5-6.2 8.7-7.5z"></path>
                <path fill="#FBBC05" d="M5.3 7c.3-1.3.7-2.6 1.3-3.7 1.7 1.3 3.3 2.5 5 3.8-.7 1.9-.7 4 0 5.8-1.7 1.3-3.3 2.5-5 3.8-1.5-2.9-2-6.4-1.3-9.7z"></path>
                <path fill="#4285F4" d="M20.3 7.2c4.8 0 9.6 0 14.4 0 .5 2.6.4 5.4-.4 8-.7 2.4-2 4.6-3.9 6.2-1.6-1.2-3.2-2.5-4.9-3.7 1.6-1.1 2.7-2.8 3.1-4.6-2.8 0-5.6 0-8.3 0 0-2 0-4 0-5.9z"></path>
                <path fill="#34A853" d="M6.6 16.7c1.7-1.3 3.3-2.5 5-3.8.6 1.8 1.9 3.5 3.5 4.6 1 .7 2.2 1.2 3.4 1.4 1.2.2 2.4.2 3.7 0 1.2-.2 2.4-.6 3.4-1.3 1.6 1.2 3.2 2.5 4.9 3.7-1.8 1.6-3.9 2.7-6.3 3.2-2.6.6-5.3.6-7.8-.1-2-.5-3.9-1.5-5.6-2.7-1.7-1.3-3.2-3-4.2-5z"></path>
              </svg></div>
          </section>
          <section class="socials_btn" style="cursor: pointer;">
            <div class="_provider-button_label__1gtgk"><span style="white-space: nowrap;">Sign in with</span> <svg viewBox="5 -5 30 30" enable-background="new 5 -5 30 30" title="facebook" class="facebook_img">
                <path fill="#3C5B9A" d="M33.3-5H6.7C5.7-5 5-4.3 5-3.3v26.7c0 .9.7 1.6 1.7 1.6H21V13.4h-3.9V8.9H21V5.6c0-3.9 2.4-6 5.8-6 1.7 0 3.1.1 3.5.2v4h-2.4c-1.9 0-2.2.9-2.2 2.2v2.9h4.5l-.6 4.5h-3.9V25h7.6c.9 0 1.7-.7 1.7-1.7V-3.3c0-1-.7-1.7-1.7-1.7z"></path>
              </svg></div>
          </section>
        </div>
        <section class="auth-form__meta">
          <p>
            Don't have an account?
            <a href="#/register" class="primary">Register now</a></p> <a href="#/forgot-password" class="primary">Forgot Password</a>
        </section>
      </div>
    </form>
  </div>
</main>
`;
